File: hering/hering.cpp
#include <iostream>
#include <cmath>

using namespace std;

int main() {
    int n;
    cin >> n;
    int prev_x, x, prev_y, y;
    int res1, res2;
    int min_x, min_y, z;
    prev_x = prev_y = 25;
    x = y = 50;

    cout << "?" << " " << prev_x << " " << y << endl << flush;
    cin >> res1;

    cout << "?" << " " << x << " " << y << endl << flush;
    cin >> res2;

    min_x = (res2 - res1 + pow(prev_x, 2) - pow(x, 2)) / (2 * prev_x - 2 * x);

    cout << "?" << " " << x << " " << prev_y << endl << flush;
    cin >> res1;

    cout << "?" << " " << x << " " << y << endl << flush;
    cin >> res2;

    min_y = (res2 - res1 + pow(prev_y,2) - pow(y, 2)) / (2 * prev_y - 2 * y);

    cout << "?" << " " << min_x << " " << min_y << endl << flush;
    cin >> z;
    z = sqrt(z);

    cout << "!" << " " << min_x << " " << min_y << " " << z << endl << flush;

    return 0;
}

File: hering/hering.py
import math
n = int(input())

prev_x = 25
x = 50

prev_y = 25
y = 50

print(f'? {prev_x} {y}')
res1 = int(input())
print(f'? {x} {y}')
res2 = int(input())

min_x = int((res2 - res1 + prev_x**2 - x**2) / ((2 * prev_x) - (2 * x)))

print(f'? {x} {prev_y}')
res1 = int(input())
print(f'? {x} {y}')
res2 = int(input())

min_y = int((res2 - res1 + prev_y**2 - y**2) / ((2 * prev_y) - (2 * y)))

print(f'? {min_x} {min_y}')
z = int(math.sqrt(int(input())))


print(f'! {min_x} {min_y} {z}')
